Best practices for backup: robocopy

core-man opened the pull request:
Related to #43.

Referring to https://core-man.github.io/blog/post/backup/#windows

core-man commented:
@zhaozhiyuan1989, could you please help improve `backup.bat`?

seisman commented:
I feel we should add a disclaimer to both `backup.sh` and `backup.bat`.

core-man commented:
> I feel we should add a disclaimer to both `backup.sh` and `backup.bat`.

That's a good idea.

seisman commented on source/best-practices/backup.rst:
```diff
@@ -111,6 +111,9 @@ Windows
 robocopy
 ^^^^^^^^
 
+使用 ``robocopy`` 命令进行备份。读者可以参考 Batch 脚本 :download:`backup.bat`\ 。点击下载后，
+修改源目录、备份目录以及想要备份的子目录。然后双击该 Batch 脚本即可直接运行，也可以打开 CMD 窗口，
+再输入 Batch 脚本名以运行脚本。
```
```suggestion
使用 `robocopy <https://docs.microsoft.com/zh-cn/windows-server/administration/windows-commands/robocopy>`__
命令进行备份。读者可以参考 Batch 脚本 :download:`backup.bat`\ 。点击下载后，
修改源目录、备份目录以及想要备份的子目录。然后双击该 Batch 脚本即可直接运行，也可以打开 CMD 窗口，
再输入 Batch 脚本名以运行脚本。
```

core-man commented on source/best-practices/backup.bat:
```diff
@@ -0,0 +1,21 @@
+REM   /mir        : 备份目录成为镜像 （复制子目录；删除备份目录下源目录中不存在的文件和目录）
+REM   /mt[:n]     : 使用 n 个线程进行多线程复制（默认值为 8）。n 至少为 1，但不得大于 128
+REM   /log+:file  : 将备份状态以追加的方式输出到日志文件 file 中
+REM 可以参考 http://www.svngo.net/article61 了解 robocopy 命令更多用法
```
```suggestion
REM 可以参考 https://docs.microsoft.com/zh-cn/windows-server/administration/windows-commands/robocopy 了解 robocopy 命令更多用法

```

seisman commented on source/best-practices/backup.rst:
```diff
@@ -80,6 +80,11 @@ rsync
    # 执行命令开始备份
    $ backup.sh
 
+.. admonition:: 免责声明
+
+   本脚本完全基于编写者在自己的计算机上备份文件的经验总结，并不具备普遍适用性，
+   仅供读者参考。
+
```
Better to move the disclaimer into the script.

core-man replied:
Done in 51a1439.

core-man commented on source/best-practices/backup.bat:
```diff
@@ -0,0 +1,22 @@
+REM 可以参考 https://docs.microsoft.com/zh-cn/windows-server/administration/windows-commands/robocopy 了解 robocopy 命令更多用法
+
+
+robocopy d:\directory1 e:\backup\directory1 /mir /mt:16 /log+:e:\backup.txt
+robocopy d:\directory2 e:\backup\directory2 /mir /mt:16 /log+:e:\backup.txt
```
```suggestion
robocopy d:\directory1 e:\backup\directory1 /mir /mt /log+:e:\backup.txt
robocopy d:\directory2 e:\backup\directory2 /mir /mt /log+:e:\backup.txt

```

core-man commented on source/best-practices/backup.bat:
```diff
@@ -0,0 +1,22 @@
+REM  最近修改日期： 2021年2月4日
+REM
+
+echo Check if the source and backup paths exist
+pause
```
Actually, I don't know what will happen if source folder, the backup folder `e:\backup` and its subfolders `e:\backup/directory1` , `e:\backup/directorys`  do not exist.

core-man replied:
> 1、执行脚本会自动创建备份目录及子目录；

Great~

> 2、如果 `source` 文件夹不存在会报错 `系统找不到文件`；

This is not a big problem. So I deleted those two lines.

> 3、如果文件夹的名字有中文需要把文件格式改为 `ANSI`；

A note is added in `backup.bat`. See   6a0f501.

> 4、如果文件夹的名字有空格需要加引号，比如 `robocopy "e:\我 测试 一下" e:\back\ /mir /mt /log+:e:\backup.txt`

I added `"` for all the folders and files. See  6a0f501.

zhaozhiyuan1989 commented:
> @zhaozhiyuan1989, could you please help test and improve `backup.bat`? You man revise both source and backup folders.

OK.

core-man commented on source/best-practices/backup.rst:
```diff
@@ -80,6 +80,11 @@ rsync
    # 执行命令开始备份
    $ backup.sh
 
+.. admonition:: 免责声明
+
+   本脚本完全基于编写者在自己的计算机上备份文件的经验总结，并不具备普遍适用性，
+   仅供读者参考。
+
```
```suggestion
```

core-man commented on source/best-practices/backup.rst:
```diff
@@ -111,6 +116,15 @@ Windows
+修改源目录、备份目录以及想要备份的子目录。然后双击该 Batch 脚本即可直接运行，也可以打开 CMD 窗口，
+再输入 Batch 脚本名以运行脚本。
+
+.. admonition:: 免责声明
+
+   本脚本完全基于编写者在自己的计算机上备份文件的经验总结，并不具备普遍适用性，
+   仅供读者参考。
```
```suggestion
```

core-man commented on source/best-practices/backup.bat:
```diff
@@ -0,0 +1,28 @@
+
+
+robocopy "d:\directory1" "e:\backup\directory1\" /mir /mt /log+:"e:\backup.txt"
+robocopy "d:\directory2" "e:\backup\directory2\" /mir /mt /log+:"e:\backup.txt"
```
As compared to `rsync`, what's the difference if there is a `\` after both source and backup folders, e.g., "d:\directory1\" and `e:\backup\directory1\`.

seisman replied:
When you connect your external hard drive to your computer, `E:` seems a reasonable drive letter.

zhaozhiyuan1989 replied:
1. 不可以在目录后加 `\`，会报错；
2. 在 Windows 中，可以直接使用移动硬盘的盘符；
3. 可以加引号；
4. 可以使用相对路径，但是不能把目标目录放到源目录下，会无限复制。

zhaozhiyuan1989 replied:
> 1. 不可以在目录后加 `\`，会报错；
> 2. 在 Windows 中，可以直接使用移动硬盘的盘符；
> 3. 可以加引号；
> 4. 可以使用相对路径，但是不能把目标目录放到源目录下，会无限复制。

第 1 点说的有问题，目录后可以加 `\`，但是和不加没有区别，之前测试报错是加**引号**引起的。
**重要的问题**，目录有空格必须加引号，但是加引号可能会造成一些无法预料的问题，比如命令无法识别出目标目录。
https://www.soinside.com/question/eQwSj6pC7gkaY6tNN2tWuY

core-man replied:
> When you connect your external hard drive to your computer, `E:` seems a reasonable drive letter.

Done in https://github.com/seismo-learn/seismology101/pull/175#discussion_r570296983.

core-man commented on source/best-practices/backup.rst:
```diff
@@ -111,6 +111,15 @@ Windows
 ^^^^^^^^
 
+使用 `robocopy <https://docs.microsoft.com/zh-cn/windows-server/administration/windows-commands/robocopy>`__
+命令进行增量备份，只同步有改动的文件。假设要备份 D 盘，移动硬盘下的备份目录为 :file:`/mnt/seismo-learn/backup/` 。
+使用以下命令可以将 D 盘里的所有子目录和文件完整同步到备份目录下，此时备份目录是 D 盘的一个镜像::
+
+    $ robocopy "d:" "/mnt/seismo-learn/backup/" /mir /mt
```
We have to change the backup path to a portable driver.

core-man commented on source/best-practices/backup.rst:
```diff
@@ -111,6 +112,15 @@ Windows
 ^^^^^^^^
 
+使用 `robocopy <https://docs.microsoft.com/zh-cn/windows-server/administration/windows-commands/robocopy>`__
+命令进行增量备份，只同步有改动的文件。假设要备份是整个 D 盘，移动硬盘下的备份目录为 :file:`/mnt/seismo-learn/backup/` 。
+使用以下命令可以将 D 盘里的所有子目录和文件完整同步到备份目录下，此时备份目录是 D 盘的一个镜像::
+
+    $ robocopy "d:" "/mnt/seismo-learn/backup/" /mir /mt
```
Does this work?

zhaozhiyuan1989 replied:
在 `robocopy`  命令中，源目录和目标目录后都不能加 `\` ，否则会报错

zhaozhiyuan1989 replied:
经测试，支持相对路径 `robocopy "..\我 测试 一下" "back2\test"  /mir /mt /log+:e:\backup.txt`

core-man replied:
can we use "d:" to indicate the while disk d?

zhaozhiyuan1989 replied:
经测试，创建日志文件时路径必须存在，否则会报错。就是说，可以创建文件，不能创建文件夹

zhaozhiyuan1989 replied:
> can we use "d:" to indicate the while disk d?

正在测试，日志文件在不断更新，移动硬盘可用空间也在变小，但是看不到复制过来的内容

core-man replied:
> 经测试，创建日志文件时路径必须存在，否则会报错。就是说，可以创建文件，不能创建文件夹

A note is added in 1ae7b894a770eb6c7d2b01db2d07a27cac6a3a14.

seisman commented on source/best-practices/backup.bat:
```diff
@@ -0,0 +1,28 @@
+robocopy "d:\directory1" "e:\backup\directory1\" /mir /mt /log+:"e:\backup.txt"
+robocopy "d:\directory2" "e:\backup\directory2\" /mir /mt /log+:"e:\backup.txt"
+
+REM 源目录和备份目录
+REM   d:\directory1 和 d:\directory2 : 要备份的源目录
+REM   e:\backup\      : 备份目录
+REM   e:\backup.txt   : 备份目录下的备份日志
+
+REM robocopy 命令常用选项：
+REM   /mir        : 备份目录成为镜像 （复制子目录；删除备份目录下源目录中不存在的文件和目录）
+REM   /mt[:n]     : 使用 n 个线程进行多线程复制（默认值为 8）。n 至少为 1，但不得大于 128
+REM   /log+:file  : 将备份状态以追加的方式输出到日志文件 file 中
+REM 可以参考 https://docs.microsoft.com/zh-cn/windows-server/administration/windows-commands/robocopy 了解 robocopy 命令更多用法
+
+REM 注意事项：
+REM   1. 如果命令中有中文，需要将脚本另存为成 ANSI 编码格式，而不是 UTF-8 格式
+REM   2. 如果文件夹的名字有空格，必须加引号
```
Move the comments before the commands.

core-man replied:
done in 2d6afa7.

seisman commented on source/best-practices/backup.bat:
```diff
@@ -0,0 +1,28 @@
+REM 可以参考 https://docs.microsoft.com/zh-cn/windows-server/administration/windows-commands/robocopy 了解 robocopy 命令更多用法
+
+REM 注意事项：
+REM   1. 如果命令中有中文，需要将脚本另存为成 ANSI 编码格式，而不是 UTF-8 格式
```
The batch script is using UTF-8 encoding, but it should use ANSI, which is the default encoding on Windows.

core-man replied:
Update the note. You mean we should use ANSI to write the script?

seisman replied:
Yes, the script is only useful for Windows users, so it makes no sense to use UTF-8.

zhaozhiyuan1989 commented on source/best-practices/backup.bat:
```diff
@@ -0,0 +1,28 @@
+REM   2. 如果文件夹的名字有空格，必须加引号
+
+robocopy "d:\directory1" "e:\backup\directory1\" /mir /mt /log+:"e:\backup.txt"
+robocopy "d:\directory2" "e:\backup\directory2\" /mir /mt /log+:"e:\backup.txt"
```
这里为啥写两行命令

core-man replied:
So that the users can choose to backup different folders in disk D.

seisman commented on source/best-practices/backup.bat:
```diff
@@ -0,0 +1,28 @@
+REM   1. 如果命令中有中文，需要将脚本另存为成 ANSI 编码格式，这也是 Windows 的默认编码格式。
+REM      本脚本是 Linux 下采用 UTF-8 格式编写的。
+REM   2. 如果文件夹的名字有空格，必须加引号
+
+robocopy "d:\directory1" "e:\backup\directory1\" /mir /mt /log+:"e:\backup.txt"
+robocopy "d:\directory2" "e:\backup\directory2\" /mir /mt /log+:"e:\backup.txt"
```
```suggestion

# 备份 d:\directory1 到 e:\backup\directory1\
robocopy "d:\directory1" "e:\backup\directory1\" /mir /mt /log+:"e:\backup.txt"

# 备份 d:\directory2 到 e:\backup\directory2\
robocopy "d:\directory2" "e:\backup\directory2\" /mir /mt /log+:"e:\backup.txt"
```

seisman commented on source/best-practices/backup.rst:
```diff
@@ -111,6 +112,15 @@ Windows
 robocopy
 ^^^^^^^^
 
+使用 `robocopy <https://docs.microsoft.com/zh-cn/windows-server/administration/windows-commands/robocopy>`__
```
```suggestion
使用 Windows 自带的 `robocopy <https://docs.microsoft.com/zh-cn/windows-server/administration/windows-commands/robocopy>`__
```

seisman commented on source/best-practices/backup.rst:
```diff
@@ -111,6 +112,15 @@ Windows
 ^^^^^^^^
 
+使用 `robocopy <https://docs.microsoft.com/zh-cn/windows-server/administration/windows-commands/robocopy>`__
+命令进行增量备份，只同步有改动的文件。假设要备份是整个 D 盘，移动硬盘下的备份目录为 :file:`/mnt/seismo-learn/backup/` 。
```
```suggestion
命令进行增量备份，只同步有改动的文件。假设要备份整个 D 盘，移动硬盘下的备份目录为 :file:`/mnt/seismo-learn/backup/` 。
```

seisman commented on source/best-practices/backup.bat:
```diff
@@ -0,0 +1,31 @@
+# 备份 d:\directory1 到 e:\backup\directory1\
+robocopy "d:\directory1" "e:\backup\directory1\" /mir /mt /log+:"e:\backup.txt"
+
+# 备份 d:\directory2 到 e:\backup\directory2\
+robocopy "d:\directory2" "e:\backup\directory2\" /mir /mt /log+:"e:\backup.txt"
```
Add `pause` at the end of the script so that the CMD window is still open when the script finishes
```suggestion
# 备份 d:\directory2 到 e:\backup\directory2\
robocopy "d:\directory2" "e:\backup\directory2\" /mir /mt /log+:"e:\backup.txt"

pause
```

core-man commented:
Some notes:

1. We use `F:\backup` to indicate the backup folder in portable driver now.
2. We want to use `D:\` to backup all the folders and files in Disk D. Zhiyuan is still testing it.
3. We remove the last `\` in the source and backup folders in `backup.bat`.
4. Still not clear what will happen if we use `""` for the source and backup folders.

seisman commented on source/best-practices/backup.bat:
```diff
@@ -0,0 +1,34 @@
+
+REM robocopy 命令常用选项：
+REM   /mir        : 备份目录成为镜像 （复制子目录；删除备份目录下源目录中不存在的文件和目录）
+REM   /mt[:n]     : 使用 n 个线程进行多线程复制（默认值为 8）。n 至少为 1，但不得大于 128
```
People may be confuse if `[]` is required.
```suggestion
REM   /mt[:n]     : 使用 n 个线程进行多线程复制（默认值为 8）。n 至少为 1，但不得大于 128。例如 /mt:4 表示 4 个线程
```

core-man commented on source/best-practices/backup.bat:
```diff
@@ -0,0 +1,35 @@
+REM robocopy 命令常用选项：
+REM   /mir        : 备份目录成为镜像 （复制子目录；删除备份目录下源目录中不存在的文件和目录）
+REM   /mt[:n]     : 使用 n 个线程进行多线程复制（默认值为 8）。n 至少为 1，但不得大于 128。
+REM                 例如 /mt 表示 8 个线程，/mt:4 表示 4 个线程
```
```suggestion
REM                 例如，/mt 表示 8 个线程，/mt:4 表示 4 个线程

```

zhaozhiyuan1989 commented:
目前备份命令执行 6 个小时，我的 D 盘总共 700G 文件，移动硬盘已经占用了 500G 的空间，但是看不到内容，备份日志文件已经 13M 了

zhaozhiyuan1989 commented:
我把备份脚本放到了 `e:\backup\` 文件夹下面进行测试（F 为移动硬盘）：
1.  `robocopy "e:"  "f:\backup\test" /mir /mt `   =>  会备份 `e:\backup\`  而不是整个硬盘
2.  `robocopy e:  f:\backup\test  /mir /mt `   =>  会备份 `e:\backup\`  而不是整个硬盘
3.  `robocopy "e:\"  "f:\backup\test " /mir /mt ` => 会报错
4.  `robocopy "e:\ "  "f:\backup\test " /mir /mt ` => 会备份整个硬盘
5.  `robocopy e:\  f:\backup\test  /mir /mt ` => 会备份整个硬盘

目前看来，如果驱动器不加 `\` 那么会备份脚本所在目录，加了 `\` 可以备份整个驱动器；
如果加了引号 `""` ，在路径最后加一个或多个空格就可以避免识别不出目标文件夹的问题

core-man commented:
> 我把备份脚本放到了 `e:\backup\` 文件夹下面进行测试（F 为移动硬盘）：
> 
> 1. `robocopy "e:"  "f:\backup\test" /mir /mt `   =>  会备份 `e:\backup\`  而不是整个硬盘
> 2. `robocopy e:  f:\backup\test  /mir /mt `   =>  会备份 `e:\backup\`  而不是整个硬盘
> 3. `robocopy "e:\"  "f:\backup\test " /mir /mt ` => 会报错
> 4. `robocopy "e:\ "  "f:\backup\test " /mir /mt ` => 会备份整个硬盘
> 5. `robocopy e:\  f:\backup\test  /mir /mt ` => 会备份整个硬盘
> 
> 目前看来，如果驱动器不加 `\` 那么会备份脚本所在目录，加了 `\` 可以备份整个驱动器；
> 如果加了引号 `""` ，在路径最后加一个或多个空格就可以避免识别不出目标文件夹的问题

Does it matter if we use upper case `E:\ ` instead of `e:\ `

core-man commented:
I have updated the script and note. Please review it.

core-man commented on source/best-practices/backup.bat:
```diff
@@ -0,0 +1,37 @@
+REM      并在目录名最后加空格 "D:\direc tory1 "）。
+REM   3. 使用 /log+:file 选项时，file（如 F:\backup\backup.txt）中的目录（即 F:\backup）必须存在
+
+# 备份 D:\directory1 到 F:\backup\directory1
```
```suggestion
REM 备份 D:\directory1 到 F:\backup\directory1
```

core-man commented on source/best-practices/backup.bat:
```diff
@@ -0,0 +1,37 @@
+# 备份 D:\directory1 到 F:\backup\directory1
+robocopy "D:\directory1 " "F:\backup\directory1 " /mir /mt /log+:"F:\backup\backup.txt"
+
+# 备份 D:\directory2 到 F:\backup\directory2
```
```suggestion
REM 备份 D:\directory2 到 F:\backup\directory2
```

core-man commented on source/best-practices/backup.rst:
```diff
@@ -111,6 +112,22 @@ Windows
+.. important::
+
+   以上命令中目录的引号、最后的斜杠和最后的空格非常重要。
+   若没有斜杠和空格（\ :file:`"D:"` 和 :file:`"F:\\backup"`\ ），备份的是脚本所在目录，并非 D 盘。
```
Since we run this command in CMD or PowerShell, I'm not sure if it will backup the current dir.

seisman commented:
I believe the script is still UTF8.

core-man commented:
> I believe the script is still UTF8.

I can only use `GBK` in VS code. See c5b57c6.

See https://www.cnblogs.com/happybirthdaytoyou/p/12239479.html

seisman commented:
This is what I see:
```
$ enca -L zh_CN *.bat
backup-ANSI.bat: Simplified Chinese National Standard; GB2312
  CRLF line terminators
backup.bat: Universal transformation format 8 bits; UTF-8
  CRLF line terminators
```

Need @zhaozhiyuan1989 to check if backup-ANSI.bat works.

zhaozhiyuan1989 commented:
> If we want to backup `e:\dir1`, I guess we also need to add whitespace inside, `"e:\dir1 "`

只有在 `源/目标` 目录加了引号且最后是 `\` 的时候才需要加空格。
比如 `"d:\test\ "` 后面必须加空格，`d:\test\` 和 `"d:\test"` 后面不用加空格，相对路径下也是如此情况。

zhaozhiyuan1989 commented:
> Does it matter if we use upper case `E:\ ` instead of `e:\ `?

不影响

zhaozhiyuan1989 commented:
今天查看了一下备份结果：
1. 看移动硬盘的占用空间应该是备份完毕了，但是看不到备份的文件，在终端使用命令可以看到文件，确认文件是隐藏了。奇怪的是，即使勾选了“显示隐藏文件”也还是看不到。和这个链接的情况类似：https://bbs.51cto.com/thread-1144509-1.html
2. 备份整个驱动器这个命令需要一些改进，因为遇到一些无法访问的目录会出错，有时候会不断循环企图备份这个目录但是又无法成功导致程序卡在这里。

zhaozhiyuan1989 commented:
目前通过执行 `robocopy d:\ f:\backup\d /mir /mt  /XD Config.Msi $RECYCLE.BIN WSLDIR /A-:H

core-man commented:
> 目前通过执行 `robocopy d:\ f:\backup\d /mir /mt /XD Config.Msi $RECYCLE.BIN WSLDIR /A-:H` 把 D 盘成功备份了。
> 
> 1. 加上 `/A-:H` 后可以正常显示备份的目录了；
> 2. 日志文件太大打不开，我每次都需要用 `tail` 来查看进度，不是很友好，我觉得在脚本中加个 `pause` 然后在终端中实时查看更好，所以在命令中去除了；
> 3. 加上 `/XD` 后跟需要排除的目录，避免了无法备份后有可能卡死的情况；
> 4. 此条命令为镜像增量备份，好处为可随时终止后重新开始不影响备份，第一次备份非常慢，但是之后备份非常快。需要注意的地方是，因为是镜像备份，在备份目录进行的修改在下次同步后会被源目录的内容覆盖，**一定要注意，执行命令时确认目标目录是否正确，避免重要数据被同步命令抹除**；

Do we have those issues if we want to backup subdir in D disk, e.g. `d:\dir1`. Like those we so in `backup.bat`.

zhaozhiyuan1989 commented:
> > 目前通过执行 `robocopy d:\ f:\backup\d /mir /mt /XD Config.Msi $RECYCLE.BIN WSLDIR /A-:H` 把 D 盘成功备份了。
> > 
> > 1. 加上 `/A-:H` 后可以正常显示备份的目录了；
> > 2. 日志文件太大打不开，我每次都需要用 `tail` 来查看进度，不是很友好，我觉得在脚本中加个 `pause` 然后在终端中实时查看更好，所以在命令中去除了；
> > 3. 加上 `/XD` 后跟需要排除的目录，避免了无法备份后有可能卡死的情况；
> > 4. 此条命令为镜像增量备份，好处为可随时终止后重新开始不影响备份，第一次备份非常慢，但是之后备份非常快。需要注意的地方是，因为是镜像备份，在备份目录进行的修改在下次同步后会被源目录的内容覆盖，**一定要注意，执行命令时确认目标目录是否正确，避免重要数据被同步命令抹除**；
> 
> Do we have those issues if we want to backup subdir in D disk, e.g. `d:\dir1`. Like those we so in `backup.bat`.

第一点，目前备份文件夹没有出现过隐藏的情况；
第二点，跟你备份的目录中文件个数有关，文件越多，日志内容越多；
第三点，跟你备份的目录有关，如果目录拒绝访问就没法备份；
第四点，通用答案

core-man commented:
@zhaozhiyuan1989, I have revised the text and script based on my understanding. Please review them once again.

Some key points:
- In the CMD or PowerShell, we now used `$ robocopy D:\ F:\backup /mir /mt /XD Config.Msi $RECYCLE.BIN WSLDIR /A-:H` to backup the whole D disk based on @zhaozhiyuan1989's test. Is it right?
- In the `backup.bat`, we removed `/log` option and added `pause` after each backup command and `echo` before, so that users can check the backup process, right? This is because `log.txt` may be too large.
- We use VS code to encode `backup.bat` in `GBK`, which is similar to `ANSI`. Please check if the `backup.bat` encoded in `GBK` works well in Windows @zhaozhiyuan1989.

core-man commented on source/best-practices/backup.rst:
```diff
@@ -111,6 +112,20 @@ Windows
+使用 Windows 自带的 `robocopy <https://docs.microsoft.com/zh-cn/windows-server/administration/windows-commands/robocopy>`__
+命令进行增量备份。这意味着只有第一次备份的时候需要花比较多的时间来同步文件，之后再使用该命令进行备份
+只会同步有改动的文件。假设要备份整个 D 盘，移动硬盘下的备份目录为 :file:`F:\\backup` 。
+使用以下命令可以将 D 盘同步到备份目录下，此时备份目录是 D 盘的一个镜像::
```
```suggestion
打开 CMD 或 PowerShell，使用以下命令可以将 D 盘同步到备份目录下，此时备份目录是 D 盘的一个镜像::
```

zhaozhiyuan1989 commented on source/best-practices/backup.bat:
```diff
@@ -1,37 +1,32 @@
-REM      并在目录名最后加空格 "D:\direc tory1 "）。
-REM   3. 使用 /log+:file 选项时，file（如 F:\backup\backup.txt）中的目录（即 F:\backup）必须存在
+REM ע�����
+REM   1. ��������������ģ��ű���Ҫ���� ANSI �����ʽ��Windows �µ�Ĭ�ϱ����ʽ����
```
这句话有些问题。
如果是右键新建文本文件，那么默认是 UTF-8；
如果是 CMD 中 `echo 1 > 1.bat` 那么默认也是 UTF-8；
只有在 CMD 中 `echo 测试 > 1.bat` 默认才会是 ANSI。

core-man replied:
> 如果是右键新建文本文件，那么默认是 UTF-8；
If you write some Chinese, will it become ANSI?
If yes, I think we can ignore this issue since the file with Chinese will be ANSI by default.

core-man commented on source/best-practices/backup.rst:
```diff
@@ -111,6 +112,20 @@ Windows
+
+.. important::
+
+   以上命令中 D 盘盘符后的斜杠非常重要，省略的话可能无法备份整个 D 盘。
```
```suggestion
   以上命令中 D 盘盘符后的反斜杠（\ :file:`D:\\`\ ）非常重要，省略的话可能无法备份整个 D 盘。
```

zhaozhiyuan1989 commented on source/best-practices/backup.rst:
```diff
@@ -113,17 +113,15 @@ robocopy
+使用以下命令可以将 D 盘同步到备份目录下，此时备份目录是 D 盘的一个镜像::
 
-    $ robocopy "D:\ " "F:\backup " /mir /mt
+    $ robocopy D:\ F:\backup /mir /mt /XD Config.Msi $RECYCLE.BIN WSLDIR /A-:H
```
这句命令中 ` /XD Config.Msi $RECYCLE.BIN WSLDIR` 后面三个文件夹，前两个是系统自己产生的（可自行搜索为何产生）无法被备份，所以添加到忽略列表，最后的 `WSLDIR` 是我 WSL 的工作目录，只要有一点更改就需要重新备份整个好几十G的 `ext4.vhdx` 文件，所以也添加到忽略列表了。
读者应该根据自己的实际情况把无法备份的文件夹或者不想备份的文件夹添加到忽略列表。

core-man replied:
If those folders do not exist, does it matter? I guess different users may have different dir names, e.g., `WSLDIR`.

core-man replied:
Updated in 670c8e5. Please check.

zhaozhiyuan1989 replied:
> If those folders do not exist, does it matter? I guess different users may have different dir names, e.g., `WSLDIR`.

没有影响

zhaozhiyuan1989 reviewed (commented):
backup.bat 脚本在我的系统上是 ANSI 格式的了。有一点建议就是在脚本中的示例可以举两个比较典型的案例，比如把第二个改成文件名为中文且有空格的情况

core-man commented:
> backup.bat 脚本在我的系统上是 ANSI 格式的了。

Great to know.

> 有一点建议就是在脚本中的示例可以举两个比较典型的案例，比如把第二个改成文件名为中文且有空格的情况

Done in 670c8e5.

zhaozhiyuan1989 commented:
还有需要验证的内容请在留言下列明，上面对话太多看着有点乱

core-man commented:
> 如果是右键新建文本文件，那么默认是 UTF-8；

If you write some Chinese, will it become ANSI?
If yes, I think we can ignore this issue since the file with Chinese will be ANSI by default.

core-man commented:
Have you ever tried to run ` $ robocopy D:\ F:\backup /mir /mt /XD Config.Msi $RECYCLE.BIN WSLDIR /A-:H` in the CMD or PowerShell directly instead of a script.

core-man commented:
Please review the text and script again to further polish and double-check it.

zhaozhiyuan1989 commented:
> @zhaozhiyuan1989, have you ever tried to run ` $ robocopy D:\ F:\backup /mir /mt /XD Config.Msi $RECYCLE.BIN WSLDIR /A-:H` in the CMD or PowerShell directly instead of a script.

测试过，效果一样

zhaozhiyuan1989 commented:
> > 这句话有些问题。
> > 如果是右键新建文本文件，那么默认是 UTF-8；
> 
> If you write some Chinese, will it become ANSI?
> If yes, I think we can ignore this issue since the file with Chinese will be ANSI by default.
> 
> > 如果是 CMD 中 `echo 1 > 1.bat` 那么默认也是 UTF-8；
> 
> If you write some Chinese into `1.bat` again, will it become ANSI? `echo 测试 >> 1.bat`
> If yes, I think we can ignore this issue since the file with Chinese will be ANSI by default.
> 
> > 只有在 CMD 中 `echo 测试 > 1.bat` 默认才会是 ANSI。

如果都是在终端上用 `echo` 进行操作，那么只要输入中文就会变成 ANSI 编码。但是绝大部分情况下，Windows 用户都是直接打开文本进行编辑的，这样即使输入中文仍然是 UTF8 编码，如果在此基础上在终端中输入中文，那么文本格式会变成 ANSI，显而易见，之前在 UTF8 编码下输入的中文会变成乱码。

core-man commented:
Move `/A-:H` ahead of `/XD` so that readers won't regard it is an exclued folder.
`$ robocopy D:\ F:\backup /mir /mt /A-:H /XD Config.Msi $RECYCLE.BIN WSLDIR`

Update the encoding note in `backup.bat`.

I'd like to merge the PR if both of you approve. @seisman @zhaozhiyuan1989

zhaozhiyuan1989 reviewed (commented):
可以参考 https://blog.csdn.net/langliu/article/details/83030664 重试参数一节，加上 `/R:10 /W:10`   减少备份文件夹失败的重试次数，节省时间。
其它没有了。

seisman commented on source/best-practices/backup.rst:
```diff
@@ -2,8 +2,9 @@
 :本节贡献者: |姚家园|\（作者）、
-             |田冬冬|\（作者）
-:最近更新日期: 2021-02-02
+             |田冬冬|\（作者）、
+             |赵志远|\（审稿）
+:最近更新日期: 2021-02-04
```
```suggestion
             |田冬冬|\（作者）、
             |赵志远|\（作者）
:最近更新日期: 2021-02-08
```

seisman commented on source/best-practices/backup.rst:
```diff
@@ -111,6 +112,24 @@ Windows
 
+使用 Windows 自带的 `robocopy <https://docs.microsoft.com/zh-cn/windows-server/administration/windows-commands/robocopy>`__
+命令进行增量备份。这意味着只有第一次备份的时候需要花比较多的时间来同步文件，之后再使用该命令进行备份
+只会同步有改动的文件。假设要备份整个 D 盘，移动硬盘下的备份目录为 :file:`F:\\backup` 。
```
Are the two backslashes necessary?

core-man replied:
Only for `ReST` grammar. Only one `\` displays  on the website.

![Screenshot from 2021-02-08 15-16-34](https://user-images.githubusercontent.com/50591376/107187246-aaae9d00-6a20-11eb-8df1-a4d56e774ac4.png)

seisman commented on source/best-practices/backup.rst:
```diff
@@ -111,6 +112,24 @@ Windows
+只会同步有改动的文件。假设要备份整个 D 盘，移动硬盘下的备份目录为 :file:`F:\\backup` 。
+打开 CMD 或 PowerShell，使用以下命令可以将 D 盘同步到备份目录下，此时备份目录是 D 盘的一个镜像::
+
+    $ robocopy D:\ F:\backup /mir /mt /A-:H /XD Config.Msi $RECYCLE.BIN WSLDIR
```
I suggest removing the `WSLDIR`, because it's created by users.
```suggestion
    $ robocopy D:\ F:\backup /mir /mt /A-:H /XD Config.Msi $RECYCLE.BIN
```

core-man replied:
@zhaozhiyuan1989，`robocopy` cannot copy WSL, right?

seisman replied:
The answer may be yes, but the directory `WSLDIR` is created by users. I think the default WSL dir is in `C:`, no?

core-man replied:
> The answer may be yes, but the directory `WSLDIR` is created by users. I think the default WSL dir is in `C:`, no?

Yes, I'd like to put it in the note if the answer is yes.

core-man replied:
Revised in e82e552.

zhaozhiyuan1989 replied:
robocopy 可以备份 WSL2，但是由于 WSL2 是一个整体的 ext4.vhdx 文件，所以只要稍微改动就会备份整个系统，所以我不建议备份它

seisman commented on source/best-practices/backup.rst:
```diff
@@ -111,6 +112,24 @@ Windows
+
+   以上命令中 D 盘盘符后的反斜杠（\ :file:`D:\\`\ ）非常重要，省略的话可能无法备份整个 D 盘。
+
+   ``/XD`` 选项后的目录（如 :file:`Config.Msi`\ 、\ :file:`$RECYCLE.BIN`\ 、\ :file:`WSLDIR`\ ）
```
```suggestion
   ``/XD`` 选项后的目录（如 :file:`Config.Msi`\ 、\ :file:`$RECYCLE.BIN`\ ）
```

core-man commented:
> 可以参考 https://blog.csdn.net/langliu/article/details/83030664 重试参数一节，加上 `/R:10 /W:10` 减少备份文件夹失败的重试次数，节省时间。
> 其它没有了。

Added in 83817cb only for `D:\` backup. The authors may have to check those parameters when they try `backup.bat`.

seisman commented on source/best-practices/backup.bat:
```diff
@@ -0,0 +1,36 @@
+REM ע�����
+REM   1. ���ű��д������ģ���Ҫ���� ANSI �����ʽ����������롣
+REM   2. ��Ŀ¼����ڿո񣬱������ţ��� "D:\�� Ŀ"
+
+echo "���� D:\directory �� F:\backup\directory"
+robocopy D:\directory F:\backup\directory /mir /mt
+pause
+
+echo "���� D:\���� �� F:\backup\����"
+robocopy D:\���� F:\backup\���� /mir /mt
+pause
+
+echo "���� D:\�� Ŀ �� F:\backup\�� Ŀ"
+robocopy "D:\�� Ŀ" "F:\backup\�� Ŀ" /mir /mt
+pause
```
Also add the `robocopy D:\ F:\backup /mir /mt /R:10 /W:10 /A-:H /XD Config.Msi $RECYCLE.BIN` command?

core-man replied:
For diectories in `D:`, we can backup them only with `/mir /mt`. Right, @zhaozhiyuan1989 ?

Thus, do we want to add `/R:10 /W:10 /A-:H /XD Config.Msi $RECYCLE.BIN`?

zhaozhiyuan1989 replied:
答案是可以。我觉得这条命令可以放在 note 里。

seisman reviewed (approved):
This PR looks good to me, although I don't read your full discussions.

zhaozhiyuan1989 commented:
这个PR我目前没什么问题了
